substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/JavaUtilZipSubstitutions.java

```diff
@@ -209,62 +228,111 @@ private int deflateBytes(long addr, byte[] b, int off, int len, int flush) {
+            return doDeflate(addr, inputPinned.addressOfArrayElement(inputOff), inputLen, WordFactory.pointer(outputBuffer), outputLen, flush, params);
+        }
     }
 
+    // (JDK 11)src/java.base/share/native/libzip/Deflater.c-Java_java_util_zip_Deflater_deflateBufferBytes(JNIEnv *env, jobject this, jlong addr,
+    //                                         jlong inputBuffer, jint inputLen,
+    //                                         jbyteArray outputArray, jint outputOff, jint outputLen,
+    //                                         jint flush, jint params)
     @Substitute
-    @TargetElement(onlyWith = JDK9OrLater.class)
-    @SuppressWarnings({"unused", "static-method"})
-    long deflateBufferBytes(long addr,
-                    long inputAddress, int inputLen,
-                    byte[] outputArray, int outputOff, int outputLen,
-                    int flush, int params) {
-        throw VMError.unsupportedFeature("JDK9OrLater: Target_java_util_zip_Deflater.deflateBufferBytes(long, long, int, byte[], int, int, int, int)");
+    @TargetElement(onlyWith = JDK11OrLater.class)
+    private long deflateBufferBytes(long addr, long inputBuffer, int inputLen, byte[] outputArray, int outputOff, int outputLen, int flush, int params) {
+        try (PinnedObject outputPinned = PinnedObject.create(outputArray)) {
+            return doDeflate(addr, WordFactory.pointer(inputBuffer), inputLen, outputPinned.addressOfArrayElement(outputOff), outputLen, flush, params);
+        }
     }
 
+    // (JDK 11)src/java.base/share/native/libzip/Deflater.c-Java_java_util_zip_Deflater_deflateBufferBuffer(JNIEnv *env, jobject this, jlong addr,
+    //                                         jlong inputBuffer, jint inputLen,
+    //                                         jlong outputBuffer, jint outputLen,
+    //                                         jint flush, jint params)
     @Substitute
-    @TargetElement(onlyWith = JDK9OrLater.class)
-    @SuppressWarnings({"unused", "static-method"})
-    long deflateBufferBuffer(long addr,
-                    long inputAddress, int inputLen,
-                    long outputAddress, int outputLen,
-                    int flush, int params) {
-        throw VMError.unsupportedFeature("JDK9OrLater: Target_java_util_zip_Deflater.deflateBufferBuffer(long, long, int, long, int, int, int)");
+    @TargetElement(onlyWith = JDK11OrLater.class)
+    private long deflateBufferBuffer(long addr, long inputBuffer, int inputLen, long outputBuffer, int outputLen, int flush, int params) {
+        return doDeflate(addr, WordFactory.pointer(inputBuffer), inputLen, WordFactory.pointer(outputBuffer), outputLen, flush, params);
     }
 
-    // Deflater.c-Java_java_util_zip_Deflater_getAdler(JNIEnv *env, jclass cls, jlong addr)
+    // (JDK 11)src/java.base/share/native/libzip/Deflater.c-ava_java_util_zip_Deflater_getAdler(JNIEnv *env, jclass cls, jlong addr)
```

Comment:
This method seems unchanged from JDK-8.  Is the `(JDK 11)` on the comment just  a notation about where code is now, as opposed to `jdk8u/jdk/src/share/native/java/util/zip/Deflater.c`?  If you are going to cite anything I would cite the JDK-8 location rather than making it look like this was a JDK-11 method.  (Even though it does not say `JDK11OrLater`.)

Reply:
Since the comments are for reference, I went with Java 11 because it will generally have the more up-to-date reference code and that's the version I looked at when I checked the implementation over.  I assume that you wanted these comments for reference/matching against the originals, not for marking what API methods come from what JDK versions.

Reply:
There is also the use of the comments for code archeologists: our implementation of the method was based on JDK-8.  The previous comments were silent on that because we never thought we would make it past JDK-8.  (And we still haven't \:-)